### ISSUE TITLE
Refactor Problem, State, and Timestepper types

### ISSUE THREAD

**Issue by glwagner** [FourierFlows/FourierFlows.jl#92]
We'd like to refactor the internal type system of FourierFlows.

Consider the current form of `Problem` and `State`:

```julia
mutable struct Problem <: AbstractProblem
  grid::AbstractGrid
  vars::AbstractVars
  params::AbstractParams
  eqn::AbstractEquation
  ts::AbstractTimeStepper
  state::AbstractState
  t::Float64
  step::Int
end
```

```julia
mutable struct State{T,dim} <: AbstractState
  t::Float64
  step::Int
  dt::Float64
  sol::Array{T,dim}
end
```

Both are mutable. Also, both redundantly store `t` and `step`, and `State` mysteriously includes `dt` among it's fields. This is not currently used, though it might be if one of the time-steppers were adaptive.

We feel that the Grid, Vars, Params, Equation, and TimeStepper types are effective. However, State should change.

One proposal is to eliminate `State` and create a new type called `Clock`, defined thusly:

```julia
mutable struct Clock{T}
  t::T
  dt::T
  step::Int
end
```

Then Problem becomes

```Julia
struct Problem{Tsol,d} 
  grid::AbstractGrid
  vars::AbstractVars
  params::AbstractParams
  eqn::AbstractEquation
  ts::AbstractTimeStepper
  clock::Clock
  sol::Array{Tsol,d}
end
```

or something of the sort. But not mutable!

This will require us to retool all of the timesteppers, the physical modules, and the docs. It's not a minor change. For example, `sol` must be passed explicitly to `stepforward!` rather than `s::State`, and a new variable `cl::Clock` must be passed as well.

It may make our lives easier down the road, however.

**Comment by glwagner** [comment]
We should also add convenience functions

```julia
getindex(p::Problem, fld) = !∈ [:t, :step] ? throw(MethodError(getindex, (p, fld)) : getfield(p.clock, fld)
```

So that we can write `prob[:t]` to get the current time. Or maybe just

```julia
time(p::Problem) = p.clock.t
```

**Comment by glwagner** [comment]
Should we also change the notation `LC`? I have always been dissatisfied with this. Should we call it `L` instead? Does it make sense if 

```julia
rhs = L*sol + N
```

?

**Comment by navidcy** [comment]
1. The inner-type of the `grid` should propagate to `sol`.
2. We should create AbstractTypes: `ImplicitTimeStepper` and `ExplicitTimeStepper`

**Comment by glwagner** [comment]
I also propose that we remove the `freq` field from `Diagnostics`. This field was intended to allow computationally expensive diagnostics to be computed once every other time-step, or once ever 10 or 100 time-steps, for example. Yet we don't use it and it complicates the code so I think we should remove it. We can always add it back if we find it might be useful.

**Comment by navidcy** [comment]
I disagree; I use it!
Perhaps there is a way around it though.

**Comment by glwagner** [comment]
Do you find that it significantly speeds up your code to set `freq > 1`?

**Comment by navidcy** [comment]
Most of the issues discussed here were resolved by PR #96. @glwagner, should we keep this open?

**Comment by glwagner** [comment]
Of course not!
